File: services/outpaint_padding.py
"""Outpaint padding helpers (mirrors ui/frontend/src/lib/outpaintPadding.ts for tests)."""

from __future__ import annotations

from typing import TypedDict

PAD_STEP = 8
QWEN_NATIVE_OUTPAINT_DIM = 1024
DEFAULT_MAX_OUTPAINT_PER_PASS = 512


class OutpaintPadding(TypedDict):
    left: int
    top: int
    right: int
    bottom: int


def snap_outpaint_padding(value: int, step: int = PAD_STEP) -> int:
    v = max(0, int(round(value)))
    if v == 0:
        return 0
    return max(step, int(round(v / step)) * step)


def has_outpaint_padding(pad: OutpaintPadding) -> bool:
    return pad["left"] > 0 or pad["top"] > 0 or pad["right"] > 0 or pad["bottom"] > 0


def _split_side_into_chunks(total: int, max_per_pass: int) -> list[int]:
    if total <= 0:
        return []
    cap = max(PAD_STEP, snap_outpaint_padding(max_per_pass))
    chunks: list[int] = []
    remaining = total
    while remaining > 0:
        if remaining <= cap:
            chunks.append(remaining)
            break
        chunks.append(cap)
        remaining -= cap
    return chunks


def split_outpaint_into_stages(
    pad: OutpaintPadding,
    max_per_pass: int = DEFAULT_MAX_OUTPAINT_PER_PASS,
) -> list[OutpaintPadding]:
    left_chunks = _split_side_into_chunks(pad["left"], max_per_pass)
    top_chunks = _split_side_into_chunks(pad["top"], max_per_pass)
    right_chunks = _split_side_into_chunks(pad["right"], max_per_pass)
    bottom_chunks = _split_side_into_chunks(pad["bottom"], max_per_pass)
    num_stages = max(
        len(left_chunks),
        len(top_chunks),
        len(right_chunks),
        len(bottom_chunks),
        1,
    )
    stages: list[OutpaintPadding] = []
    for i in range(num_stages):
        stage: OutpaintPadding = {
            "left": left_chunks[i] if i < len(left_chunks) else 0,
            "top": top_chunks[i] if i < len(top_chunks) else 0,
            "right": right_chunks[i] if i < len(right_chunks) else 0,
            "bottom": bottom_chunks[i] if i < len(bottom_chunks) else 0,
        }
        if has_outpaint_padding(stage):
            stages.append(stage)
    return stages if stages else [{"left": 0, "top": 0, "right": 0, "bottom": 0}]

File: ui/frontend/src/lib/outpaintPadding.ts
export type OutpaintBoxPx = {
  bx0: number;
  by0: number;
  bx1: number;
  by1: number;
};

export type OutpaintPaddingPx = {
  left: number;
  top: number;
  right: number;
  bottom: number;
};

export type OutpaintStage = OutpaintPaddingPx;

/** Qwen model native working width (workflow scales input to this). */
export const QWEN_NATIVE_OUTPAINT_DIM = 1024;

/** Default max padding per pass per side = half the native dim. */
export const DEFAULT_MAX_OUTPAINT_PER_PASS = 512;

export const MIN_MAX_OUTPAINT_PER_PASS = 128;
export const MAX_MAX_OUTPAINT_PER_PASS = 1024;
export const MAX_OUTPAINT_PER_PASS_STEP = 64;

export const OUTPAINT_PAD_STEP = 8;

export function snapOutpaintPadding(value: number, step = OUTPAINT_PAD_STEP): number {
  const v = Math.max(0, Math.round(value));
  if (v === 0) return 0;
  return Math.max(step, Math.round(v / step) * step);
}

/** Map output canvas box (natural px) to ImagePadForOutpaint padding. Image occupies [0,w]×[0,h]. */
export function paddingFromOutpaintBox(
  box: OutpaintBoxPx,
  nat: { w: number; h: number }
): OutpaintPaddingPx {
  return {
    left: snapOutpaintPadding(Math.max(0, -box.bx0)),
    top: snapOutpaintPadding(Math.max(0, -box.by0)),
    right: snapOutpaintPadding(Math.max(0, box.bx1 - nat.w)),
    bottom: snapOutpaintPadding(Math.max(0, box.by1 - nat.h)),
  };
}

export function hasOutpaintPadding(pad: OutpaintPaddingPx): boolean {
  return pad.left > 0 || pad.top > 0 || pad.right > 0 || pad.bottom > 0;
}

export function initialOutpaintBox(nat: { w: number; h: number }): OutpaintBoxPx {
  return { bx0: 0, by0: 0, bx1: nat.w, by1: nat.h };
}

function splitSideIntoChunks(total: number, maxPerPass: number): number[] {
  if (total <= 0) return [];
  const cap = Math.max(OUTPAINT_PAD_STEP, snapOutpaintPadding(maxPerPass));
  const chunks: number[] = [];
  let remaining = total;
  while (remaining > 0) {
    if (remaining <= cap) {
      chunks.push(remaining);
      break;
    }
    chunks.push(cap);
    remaining -= cap;
  }
  return chunks;
}

/** Split total padding into stages capped by maxPerPass per side (Qwen native dim / 2 by default). */
export function splitOutpaintIntoStages(
  pad: OutpaintPaddingPx,
  maxPerPass: number = DEFAULT_MAX_OUTPAINT_PER_PASS
): OutpaintStage[] {
  const leftChunks = splitSideIntoChunks(pad.left, maxPerPass);
  const topChunks = splitSideIntoChunks(pad.top, maxPerPass);
  const rightChunks = splitSideIntoChunks(pad.right, maxPerPass);
  const bottomChunks = splitSideIntoChunks(pad.bottom, maxPerPass);
  const numStages = Math.max(
    leftChunks.length,
    topChunks.length,
    rightChunks.length,
    bottomChunks.length,
    1
  );
  const stages: OutpaintStage[] = [];
  for (let i = 0; i < numStages; i++) {
    const stage: OutpaintStage = {
      left: leftChunks[i] ?? 0,
      top: topChunks[i] ?? 0,
      right: rightChunks[i] ?? 0,
      bottom: bottomChunks[i] ?? 0,
    };
    if (hasOutpaintPadding(stage)) stages.push(stage);
  }
  return stages.length ? stages : [{ left: 0, top: 0, right: 0, bottom: 0 }];
}

export function clampMaxOutpaintPerPass(value: number): number {
  const snapped = Math.round(value / MAX_OUTPAINT_PER_PASS_STEP) * MAX_OUTPAINT_PER_PASS_STEP;
  return Math.min(MAX_MAX_OUTPAINT_PER_PASS, Math.max(MIN_MAX_OUTPAINT_PER_PASS, snapped));
}
